File: src/test/java/me/nasukhov/TukitaLearner/bot/command/LearnWordTest.java
package me.nasukhov.TukitaLearner.bot.command;

import me.nasukhov.TukitaLearner.bot.io.Channel;
import me.nasukhov.TukitaLearner.bot.io.Input;
import me.nasukhov.TukitaLearner.bot.io.Output;
import me.nasukhov.TukitaLearner.bot.io.User;
import me.nasukhov.TukitaLearner.dictionary.DictionaryRepository;
import me.nasukhov.TukitaLearner.dictionary.Word;
import me.nasukhov.TukitaLearner.study.Group;
import me.nasukhov.TukitaLearner.study.LearnedResourceRepository;
import me.nasukhov.TukitaLearner.study.ProgressTracker;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;

import static org.mockito.Mockito.*;

@SpringBootTest
public class LearnWordTest {
    @Autowired
    private LearnWord handler;

    @Autowired
    private DictionaryRepository dictionary;

    @Autowired
    private ProgressTracker progressTracker;

    @Autowired
    private LearnedResourceRepository learnedResourceRepository;

    private Output output;

    @BeforeEach
    void setup() {
        output = mock(Output.class);

        dictionary.saveAll(List.of(
            new Word("гьой", "собака", "Делает кусь"),
            new Word("кету", "кошка", "Для глажки"),
            new Word("рухья", "дерево", "Вырабатывает кислород ми спасает от жары")
        ));
    }

    @AfterEach
    void reset() {
        learnedResourceRepository.deleteAll();
        dictionary.deleteAll();
    }

    @Test
    void handleUnsupportedInput() {
        var channel = new Channel("");
        var user = new User("SomeId", "SomeName");
        Input input = new Input("learn", channel, user);

        handler.handle(input, output);

        verifyNoInteractions(output);
    }

    @Test
    void handleWhenAllWordsAreLearned() {
        var channel = new Channel("");
        var user = new User("SomeId", "SomeName");
        Input input = new Input("/learn", channel, user);

        progressTracker.setLastLearnedWords(new Group(channel.id), dictionary.findAll());

        handler.handle(input, output);

        verify(output).write("Вы изучили все слова из нашего словаря - больше новых слов нет.");
        verifyNoMoreInteractions(output);
    }

    @Test
    void handleNormally() {
        var channel = new Channel("");
        var user = new User("SomeId", "SomeName");
        Input input = new Input("/learn", channel, user);

        handler.handle(input, output);

        verify(output).write("""
                гьой - собака
                
                кету - кошка
                
                рухья - дерево"""
        );
    }
}

File: src/test/kotlin/me/nasukhov/tukitalearner/bot/command/LearnWordTest.kt
package me.nasukhov.tukitalearner.bot.command

import me.nasukhov.tukitalearner.bot.io.Channel
import me.nasukhov.tukitalearner.bot.io.Input
import me.nasukhov.tukitalearner.bot.io.Output
import me.nasukhov.tukitalearner.bot.io.User
import me.nasukhov.tukitalearner.dictionary.DictionaryEntry
import me.nasukhov.tukitalearner.dictionary.DictionaryRepository
import me.nasukhov.tukitalearner.study.Group
import me.nasukhov.tukitalearner.study.LearnedResourceRepository
import me.nasukhov.tukitalearner.study.ProgressTracker
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@Transactional
class LearnWordTest {
    @Autowired
    private lateinit var handler: LearnWord

    @Autowired
    private lateinit var dictionary: DictionaryRepository

    @Autowired
    private lateinit var progressTracker: ProgressTracker

    @Autowired
    private lateinit var learnedResourceRepository: LearnedResourceRepository

    private lateinit var output: Output

    @BeforeEach
    fun setup() {
        output = Mockito.mock(Output::class.java)

        dictionary.saveAll(
            listOf(
                DictionaryEntry("гьой", "собака", "Делает кусь"),
                DictionaryEntry("кету", "кошка", "Для глажки"),
                DictionaryEntry("рухья", "дерево", "Вырабатывает кислород ми спасает от жары"),
            ),
        )
    }

    @AfterEach
    fun reset() {
        learnedResourceRepository.deleteAll()
        dictionary.deleteAll()
    }

    @Test
    fun handleUnsupportedInput() {
        val channel = Channel("")
        val user = User("SomeId", "SomeName")
        val input = Input("learn", channel, user)

        handler.handle(input, output)

        Mockito.verifyNoInteractions(output)
    }

    @Test
    fun handleWhenAllWordsAreLearned() {
        val channel = Channel("")
        val user = User("SomeId", "SomeName")
        val input = Input("/learn", channel, user)

        progressTracker.setLastLearnedWords(Group(channel.id), dictionary.findAll())

        handler.handle(input, output)

        Mockito.verify(output).write("Вы изучили все слова из нашего словаря - больше новых слов нет.")
        Mockito.verifyNoMoreInteractions(output)
    }

    @Test
    fun handleNormally() {
        val channel = Channel("")
        val user = User("SomeId", "SomeName")
        val input = Input("/learn", channel, user)

        handler.handle(input, output)

        Mockito.verify(output).write(
            """
            гьой - собака
            
            кету - кошка
            
            рухья - дерево
            """.trimIndent(),
        )
    }
}
